optimizely/odp/lru_cache.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2022, Optimizely
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from __future__ import annotations
+from dataclasses import dataclass, field
+import threading
+from time import time
+from collections import OrderedDict
+from typing import Optional, Generic, TypeVar, Hashable
+from sys import version_info
+
+if version_info < (3, 8):
+    from typing_extensions import Protocol
+else:
+    from typing import Protocol  # type: ignore
+
+
+# generic type definitions for LRUCache parameters
+K = TypeVar('K', bound=Hashable, contravariant=True)
+V = TypeVar('V')
+
+
+class LRUCache(Generic[K, V]):
+    """Least Recently Used cache that invalidates entries older than the timeout."""
+
+    def __init__(self, capacity: int, timeout_in_secs: float):
+        self.lock = threading.Lock()
+        self.map: OrderedDict[K, CacheElement[V]] = OrderedDict()
+        self.capacity = capacity
+        self.timeout = timeout_in_secs
+
+    def lookup(self, key: K) -> Optional[V]:
```

Comment:
Just a note that Swift deals with the cache queue in synch manner here. I'm not sure why tbh, because the "save" is async. Just a note for us to think about why that might be, if wee need to be aware of this.